src/java.base/share/classes/sun/security/util/PBEUtil.java

```diff
@@ -77,116 +122,84 @@ public IvParameterSpec getIvSpec() {
+            } catch (InvalidAlgorithmParameterException iape) {
+                // should never happen
+                throw new RuntimeException("Error initializing PBES2Params");
             }
             return params;
         }
 
         /*
-         * Obtain a PBEKeySpec for Cipher services, after key and parameters
-         * validation, random generation or assignment from defaults. Used by
-         * PBES2Core (SunJCE) and P11PBECipher (SunPKCS11).
+         * Initialize PBES2Params and obtain a PBEKeySpec for Cipher services.
+         * Data from the key, parameters, defaults or random may be used for
+         * initialization. Used by PBES2Core (SunJCE) and P11PBECipher
+         * (SunPKCS11).
          */
         public PBEKeySpec getPBEKeySpec(int blkSize, int keyLength, int opmode,
                 Key key, AlgorithmParameterSpec params, SecureRandom random)
                 throws InvalidKeyException, InvalidAlgorithmParameterException {
             if (key == null) {
                 throw new InvalidKeyException("Null key");
             }
-
-            char[] passwdChars = null;
-            salt = null;
-            iCount = 0;
-            ivSpec = null;
             PBEKeySpec pbeSpec;
             byte[] passwdBytes;
+            char[] passwdChars = null;
             if (!(key.getAlgorithm().regionMatches(true, 0, "PBE", 0, 3)) ||
                     (passwdBytes = key.getEncoded()) == null) {
                 throw new InvalidKeyException("Missing password");
             }
             try {
-                boolean doEncrypt = ((opmode == Cipher.ENCRYPT_MODE) ||
-                            (opmode == Cipher.WRAP_MODE));
-
+                int iCountInit;
+                byte[] saltInit;
                 // Extract from the supplied PBE params, if present
                 if (params instanceof PBEParameterSpec pbeParams) {
                     // salt should be non-null per PBEParameterSpec
-                    salt = check(pbeParams.getSalt());
-                    iCount = check(pbeParams.getIterationCount());
-                    AlgorithmParameterSpec ivParams =
-                            pbeParams.getParameterSpec();
-                    if (ivParams instanceof IvParameterSpec iv) {
-                        ivSpec = iv;
-                    } else if (ivParams == null && doEncrypt) {
-                        // generate random IV
-                        byte[] ivBytes = new byte[blkSize];
-                        random.nextBytes(ivBytes);
-                        ivSpec = new IvParameterSpec(ivBytes);
-                    } else {
-                        throw new InvalidAlgorithmParameterException(
-                                "Wrong parameter type: IV expected");
-                    }
-                } else if (params == null && doEncrypt) {
+                    iCountInit = check(pbeParams.getIterationCount());
+                    saltInit = check(pbeParams.getSalt());
```

Comment:
Why not override params with the IvParameterSpec inside the PBEParameterSpec here? Then the PBES2Params.initialize(...) method does not need to handle PBEParameterSpec type? Seems more consistent this way.

Reply:
@valeriepeng: I agree, the only downside I see is that we'll need to do the same `pbeParams.getParameterSpec()` extraction before [this `PBES2Params.initialize()` call](https://github.com/openjdk/jdk/blob/1c596e147dac5102b31a946c905843b9b19b428e/src/jdk.crypto.cryptoki/share/classes/sun/security/pkcs11/P11PBECipher.java#L144-L146). But for this case, it's clearly more consistent to have the extraction here. We'll address this in the next iteration.

Reply:
I'm okay with this change. While I don't like extracting in two places (on the caller side) as @franferrax  said, perhaps it makes easier to understand why the iteration count and salt are not obtained from params in PBES2Params::initialize.